### DISCUSSION TITLE
Experiment with highway

### DISCUSSION THREAD

**Discussion by pca006132** [elalish/manifold#364] [Ideas]
I finally have more time now, will investigate the use of [highway](https://github.com/google/highway) to optimize sorting, which is often the most time consuming operation.

I'm thinking about modifying the memory layout for SparseIndices to make it int64 array, so it can be manipulated directly with SIMD. It seems to me that there is no solid reason to use two separate arrays, apart from making other functions a bit simpler.

**Comment by elalish** [comment]
That looks interesting! I think I mostly used 32-bit because GPUs haven't been very friendly to 64 bit, but that may have changed. I never looked into it very closely regarding timing.

**Comment by pca006132** [comment]
Not sure if 64 bit is good for the GPU, but we can always just split the 64 bit integer into two 32 bit integers, so this should be fine.

**Comment by pca006132** [comment]
@elalish I have some question about the following code:
https://github.com/elalish/manifold/blob/754a4811a5354415148ba247c2e1002c9a247c6d/src/manifold/src/boolean3.cpp#L490-L504

More specifically, for line 496, if we do `sort_by_key` with `p0q2.begin(reverse), p0q2.end(reverse)`, I think the `p` vector will be updated while the `q` vector remains unchanged? Is this intended?

**Comment by elalish** [comment]
I believe that's correct. That should have a comment - basically the p0q2 vector is unusable after this function runs - that's fine because it's not used again after being passed in, but agreed that's pretty surprising when reading the function.

**Comment by pca006132** [comment]
I did some initial experiments, it seems that we are not spending too much time in sorting the sparse indices, so the improvement is modest for small models to negative when executed in wasm. For the time consuming parts, it seems that we have to write our own sorting functions or wait for someone to come up with SIMD implementation of `sort_by_key`.

I feel that it makes more sense to tune the `autoPolicy` value better, perhaps make an autotuning script and let users try and tell use the results.
